README.md
# serve-favicon

*Node.js middleware for serving a favicon*

this scoped version forked from [expressjs/serve-favicon](https://github.com/expressjs/serve-favicon) to:
  - rewrite as an ECMAScript Module
  - remove dependencies

A favicon is a visual cue that client software, like browsers, use to identify
a site. For an example and more information, please visit
[the Wikipedia article on favicons](https://en.wikipedia.org/wiki/Favicon).

Why use this module?

  - User agents request `favicon.ico` frequently and indiscriminately, so you
    may wish to exclude these requests from your logs by using this middleware
    before your logger middleware.
  - This module caches the icon in memory to improve performance by skipping
    disk access.
  - This module provides an `ETag` based on the contents of the icon, rather
    than file system properties.
  - This module will serve with the most compatible `Content-Type`.

**Note** This module is exclusively for serving the “default, implicit favicon”,
which is `GET /favicon.ico`. For additional vendor-specific icons that require
HTML markup, additional middleware is required to serve the relevant files, for
example [serve-static](https://npmjs.org/package/serve-static).

## Install

This is an ECMAScript Module available through the
[npm registry](https://www.npmjs.com/). Installation is done using the
[`npm install` command](https://docs.npmjs.com/getting-started/installing-npm-packages-locally):

```shell
$ npm install @robireton/serve-favicon
```

## API

### favicon(path, options)

Create new middleware to serve a favicon from the given `path` to a favicon file.
`path` may also be a `Buffer` of the icon to serve.

#### Options

Serve favicon accepts these properties in the options object.

##### maxAge

The `cache-control` `max-age` directive in `ms`, defaulting to 1 year.

## Example

Typically this middleware will come very early in your stack (maybe even first)
to avoid processing any other middleware if we already know the request is for
`/favicon.ico`.

### express

```javascript
import { resolve } from 'node:path'
import express from 'express'
import favicon from '@robireton/serve-favicon'

const app = express()
app.use(favicon(resolve('static/favicon.ico')))

// Add your routes here, etc.

app.listen(3000)
```

## License

[MIT](LICENSE)


HISTORY.md
3.0.1 / 2023-01-25
==================

  * switch to @robireton dependencies

3.0.0 / 2023-01-24
==================

  * rewrite as ESM
  * `maxAge` option only accepts integer milliseconds
  * remove dependencies on `ms`, `parseurl`, `safe-buffer`

2.5.0 / 2018-03-29
==================

  * Ignore requests without `url` property
  * deps: ms@2.1.1
    - Add `week`
    - Add `w`

2.4.5 / 2017-09-26
==================

  * deps: etag@~1.8.1
    - perf: replace regular expression with substring
  * deps: fresh@0.5.2
    - Fix regression matching multiple ETags in `If-None-Match`
    - perf: improve `If-None-Match` token parsing

2.4.4 / 2017-09-11
==================

  * deps: fresh@0.5.1
    - Fix handling of modified headers with invalid dates
    - perf: improve ETag match loop
  * deps: parseurl@~1.3.2
    - perf: reduce overhead for full URLs
    - perf: unroll the "fast-path" `RegExp`
  * deps: safe-buffer@5.1.1

2.4.3 / 2017-05-16
==================

  * Use `safe-buffer` for improved Buffer API
  * deps: ms@2.0.0

2.4.2 / 2017-03-24
==================

  * deps: ms@1.0.0

2.4.1 / 2017-02-27
==================

  * Remove usage of `res._headers` private field
  * deps: fresh@0.5.0
    - Fix incorrect result when `If-None-Match` has both `*` and ETags
    - Fix weak `ETag` matching to match spec
    - perf: skip checking modified time if ETag check failed
    - perf: skip parsing `If-None-Match` when no `ETag` header
    - perf: use `Date.parse` instead of `new Date`

2.4.0 / 2017-02-19
==================

  * deps: etag@~1.8.0
    - Use SHA1 instead of MD5 for ETag hashing
    - Works with FIPS 140-2 OpenSSL configuration
  * deps: fresh@0.4.0
    - Fix false detection of `no-cache` request directive
    - perf: enable strict mode
    - perf: hoist regular expressions
    - perf: remove duplicate conditional
    - perf: remove unnecessary boolean coercions
  * perf: simplify initial argument checking

2.3.2 / 2016-11-16
==================

  * deps: ms@0.7.2

2.3.1 / 2016-01-23
==================

  * deps: parseurl@~1.3.1
    - perf: enable strict mode

2.3.0 / 2015-06-13
==================

  * Send non-chunked response for `OPTIONS`
  * deps: etag@~1.7.0
    - Always include entity length in ETags for hash length extensions
    - Generate non-Stats ETags using MD5 only (no longer CRC32)
    - Remove base64 padding in ETags to shorten
  * deps: fresh@0.3.0
    - Add weak `ETag` matching support
  * perf: enable strict mode
  * perf: remove argument reassignment
  * perf: remove bitwise operations

2.2.1 / 2015-05-14
==================

  * deps: etag@~1.6.0
   - Improve support for JXcore
   - Support "fake" stats objects in environments without `fs`
  * deps: ms@0.7.1
    - Prevent extraordinarily long inputs

2.2.0 / 2014-12-18
==================

  * Support query string in the URL
  * deps: etag@~1.5.1
    - deps: crc@3.2.1
  * deps: ms@0.7.0
    - Add `milliseconds`
    - Add `msecs`
    - Add `secs`
    - Add `mins`
    - Add `hrs`
    - Add `yrs`

2.1.7 / 2014-11-19
==================

  * Avoid errors from enumerables on `Object.prototype`

2.1.6 / 2014-10-16
==================

  * deps: etag@~1.5.0

2.1.5 / 2014-09-24
==================

  * deps: etag@~1.4.0

2.1.4 / 2014-09-15
==================

  * Fix content headers being sent in 304 response
  * deps: etag@~1.3.1
    - Improve ETag generation speed

2.1.3 / 2014-09-07
==================

  * deps: fresh@0.2.4

2.1.2 / 2014-09-05
==================

  * deps: etag@~1.3.0
    - Improve ETag generation speed

2.1.1 / 2014-08-25
==================

  * Fix `ms` to be listed as a dependency

2.1.0 / 2014-08-24
==================

  * Accept string for `maxAge` (converted by `ms`)
  * Use `etag` to generate `ETag` header

2.0.1 / 2014-06-05
==================

  * Reduce byte size of `ETag` header

2.0.0 / 2014-05-02
==================

  * `path` argument is required; there is no default icon.
  * Accept `Buffer` of icon as first argument.
  * Non-GET and HEAD requests are denied.
  * Send valid max-age value
  * Support conditional requests
  * Support max-age=0
  * Support OPTIONS method
  * Throw if `path` argument is directory.

1.0.2 / 2014-03-16
==================

  * Fixed content of default icon.

1.0.1 / 2014-03-11
==================

  * Fixed path to default icon.

1.0.0 / 2014-02-15
==================

  * Initial release


test/test.js
import assert from 'node:assert'
import http from 'node:http'
// import { describe, it, before, beforeEach, afterEach } from 'node:test'
import { cwd } from 'node:process'
import { describe, it, before } from 'node:test'
import { Buffer } from 'node:buffer'
import { join } from 'node:path'
import favicon from '../index.js'
// import { TempIcon } from './support/tempIcon.js'

const FIXTURES_PATH = join(cwd(), 'test', 'fixtures')
const ICON_PATH = join(FIXTURES_PATH, 'favicon.ico')

describe('favicon()', () => {
  describe('arguments', function () {
    describe('path', function () {
      it('should accept file path', function () {
        assert.doesNotThrow(favicon.bind(null, ICON_PATH))
      })

      it('should accept buffer', function () {
        assert.doesNotThrow(favicon.bind(null, Buffer.alloc(20)))
      })

      it('should exist', function () {
        assert.throws(favicon.bind(null, join(FIXTURES_PATH, 'nothing')), /ENOENT:.*nothing/)
      })

      it('should not be number', function () {
        assert.throws(favicon.bind(null, 12), /must be path to favicon or buffer/)
      })
    })

    describe('options.maxAge', () => {
      it('should be in cache-control', done => {
        createServer(ICON_PATH, { maxAge: 5000 }, (err, server) => {
          if (!err) {
            http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
              const headers = parseRawHeaders(res.rawHeaders)
              assert.ok(res.statusCode >= 200 && res.statusCode < 300)
              assert.strictEqual(headers['Cache-Control'], 'public, max-age=5')
              closeServer(server, () => done())
            })
          }
        })
      })

      it('should have a default', done => {
        createServer(undefined, undefined, (err, server) => {
          if (!err) {
            http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
              const headers = parseRawHeaders(res.rawHeaders)
              assert.ok(res.statusCode >= 200 && res.statusCode < 300)
              assert.match(headers['Cache-Control'], /public, max-age=[0-9]+/)
              closeServer(server, () => done())
            })
          }
        })
      })

      it('should accept 0', done => {
        createServer(ICON_PATH, { maxAge: 0 }, (err, server) => {
          if (!err) {
            http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
              const headers = parseRawHeaders(res.rawHeaders)
              assert.ok(res.statusCode >= 200 && res.statusCode < 300)
              assert.match(headers['Cache-Control'], /public, max-age=[0-9]+/)
              closeServer(server, () => done())
            })
          }
        })
      })
    })

    it('should be valid delta-seconds', done => {
      createServer(ICON_PATH, { maxAge: 1234 }, (err, server) => {
        if (!err) {
          http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
            const headers = parseRawHeaders(res.rawHeaders)
            assert.ok(res.statusCode >= 200 && res.statusCode < 300)
            assert.strictEqual(headers['Cache-Control'], 'public, max-age=1')
            closeServer(server, () => done())
          })
        }
      })
    })

    it('should floor at 0', done => {
      createServer(ICON_PATH, { maxAge: -4000 }, (err, server) => {
        if (!err) {
          http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
            const headers = parseRawHeaders(res.rawHeaders)
            assert.ok(res.statusCode >= 200 && res.statusCode < 300)
            assert.strictEqual(headers['Cache-Control'], 'public, max-age=0')
            closeServer(server, () => done())
          })
        }
      })
    })

    it('should ceil at 1 year', done => {
      createServer(ICON_PATH, { maxAge: 900000000000 }, (err, server) => {
        if (!err) {
          http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
            const headers = parseRawHeaders(res.rawHeaders)
            assert.ok(res.statusCode >= 200 && res.statusCode < 300)
            assert.strictEqual(headers['Cache-Control'], 'public, max-age=31536000')
            closeServer(server, () => done())
          })
        }
      })
    })

    it('should ceil at 1 year', done => {
      createServer(ICON_PATH, { maxAge: 900000000000 }, (err, server) => {
        if (!err) {
          http.get(`http://[::1]:${server.address().port}/favicon.ico`, res => {
            const headers = parseRawHeaders(res.rawHeaders)
            assert.ok(res.statusCode >= 200 && res.statusCode < 300)
            assert.strictEqual(headers['Cache-Control'], 'public, max-age=31536000')
            closeServer(server, () => done())
          })
        }
      })
    })

    /*
    describe('options.maxAge', function () {

      it('should accept Inifnity', function (done) {
        const server = createServer(null, { maxAge: Infinity })
        request(server)
          .get('/favicon.ico')
          .expect('Cache-Control', 'public, max-age=31536000')
          .expect(200, done)
      })
    })
    */
  })

  describe('requests', function () {
    before(function () {
      this.server = createServer()
    })
    /*
    it('should serve icon', function (done) {
      request(this.server)
        .get('/favicon.ico')
        .expect('Content-Type', 'image/x-icon')
        .expect(200, done)
    })

    it('should include cache-control', function (done) {
      request(this.server)
        .get('/favicon.ico')
        .expect('Cache-Control', /public/)
        .expect(200, done)
    })

    it('should include strong etag', function (done) {
      request(this.server)
        .get('/favicon.ico')
        .expect('ETag', /^"[^"]+"$/)
        .expect(200, done)
    })

    it('should deny POST', function (done) {
      request(this.server)
        .post('/favicon.ico')
        .expect('Allow', 'GET, HEAD, OPTIONS')
        .expect(405, done)
    })

    it('should understand OPTIONS', function (done) {
      request(this.server)
        .options('/favicon.ico')
        .expect('Allow', 'GET, HEAD, OPTIONS')
        .expect(200, done)
    })

    it('should 304 when If-None-Match matches', function (done) {
      const server = this.server
      request(server)
        .get('/favicon.ico')
        .expect(200, function (err, res) {
          if (err) return done(err)
          request(server)
            .get('/favicon.ico')
            .set('If-None-Match', res.headers.etag)
            .expect(304, done)
        })
    })

    it('should 304 when If-None-Match matches weakly', function (done) {
      const server = this.server
      request(server)
        .get('/favicon.ico')
        .expect(200, function (err, res) {
          if (err) return done(err)
          request(server)
            .get('/favicon.ico')
            .set('If-None-Match', 'W/' + res.headers.etag)
            .expect(304, done)
        })
    })

    it('should ignore non-favicon requests', function (done) {
      request(this.server)
        .get('/')
        .expect(404, 'oops', done)
    })

    it('should work with query string', function (done) {
      request(this.server)
        .get('/favicon.ico?v=1')
        .expect('Content-Type', 'image/x-icon')
        .expect(200, done)
    })
    describe('missing req.url', function () {
      it('should ignore the request', function (done) {
        const fn = favicon(ICON_PATH)
        fn({}, {}, done)
      })
    })
    */
  })

  /*
  describe('icon', function () {
    describe('file', function () {
      beforeEach(function () {
        this.icon = new TempIcon()
        this.icon.writeSync()
      })

      afterEach(function () {
        this.icon.unlinkSync()
        this.icon = undefined
      })

      it('should be read on first request', function (done) {
        const icon = this.icon
        const server = createServer(icon.path)

        request(server)
          .get('/favicon.ico')
          .expect(200, icon.data, done)
      })

      it('should cache for second request', function (done) {
        const icon = this.icon
        const server = createServer(icon.path)

        request(server)
          .get('/favicon.ico')
          .expect(200, icon.data, function (err) {
            if (err) return done(err)
            icon.unlinkSync()
            request(server)
              .get('/favicon.ico')
              .expect(200, icon.data, done)
          })
      })
    })

    describe('file error', function () {
      beforeEach(function () {
        this.icon = new TempIcon()
        this.icon.writeSync()
      })

      afterEach(function () {
        this.icon.unlinkSync()
        this.icon = undefined
      })

      it('should next() file read errors', function (done) {
        const icon = this.icon
        const server = createServer(icon.path)

        icon.unlinkSync()
        request(server)
          .get('/favicon.ico')
          .expect(500, /ENOENT/, done)
      })

      it('should retry reading file after error', function (done) {
        const icon = this.icon
        const server = createServer(icon.path)

        icon.unlinkSync()
        request(server)
          .get('/favicon.ico')
          .expect(500, /ENOENT/, function (err) {
            if (err) return done(err)
            icon.writeSync()
            request(server)
              .get('/favicon.ico')
              .expect(200, icon.data, done)
          })
      })
    })

    describe('buffer', function () {
      it('should be served from buffer', function (done) {
        const buffer = Buffer.alloc(20, '#')
        const server = createServer(buffer)

        request(server)
          .get('/favicon.ico')
          .expect('Content-Length', '20')
          .expect(200, buffer, done)
      })

      it('should be copied', function (done) {
        const buffer = Buffer.alloc(20, '#')
        const server = createServer(buffer)

        assert.equal(buffer.toString(), '####################')
        buffer.fill('?')
        assert.equal(buffer.toString(), '????????????????????')

        request(server)
          .get('/favicon.ico')
          .expect('Content-Length', '20')
          .expect(200, Buffer.from('####################'), done)
      })
    })
  })
  */
})

function createServer (path = ICON_PATH, opts = {}, next) {
  try {
    const middleware = favicon(path, opts)
    const server = http.createServer((request, response) => {
      request.path = request.url
      middleware(request, response, err => {
        response.statusCode = err ? (err.status || 500) : 404
        response.end(err ? err.message : 'oops')
      })
    })
    server.listen(() => {
      next(null, server)
    })
  } catch (err) {
    next(err, null)
  }
}

function closeServer (server, next) {
  server.closeAllConnections()
  server.close(err => {
    if (err) {
      next(err)
    } else {
      console.log('server closed')
      next()
    }
  })
}

function parseRawHeaders (raw) {
  const headers = {}
  for (let i = 0; i < (raw.length - 1); i += 2) {
    headers[raw[i]] = raw[i + 1]
  }
  return headers
}


test/support/tempIcon.js
import crypto from 'node:crypto'
import fs from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

export default class TempIcon {
  constructor () {
    this.data = crypto.pseudoRandomBytes(100)
    this.exists = false
    this.path = join(tmpdir(), `${Math.random().toString(36).slice(2)}${Math.random().toString(36).slice(2)}`)
  }

  unlinkSync () {
    if (this.exists) {
      this.exists = false
      fs.unlinkSync(this.path)
    }
  }

  writeSync () {
    if (this.exists) {
      throw new Error('already written')
    } else {
      fs.writeFileSync(this.path, this.data)
      this.exists = true
    }
  }
}


index.js
import { Buffer } from 'node:buffer'
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import etag from '@robireton/etag'
import fresh from '@robireton/fresh'

/**
 * Serves the favicon located by the given `path`.
*
* @public
* @param {String|Buffer} resource
* @param {Object} [options]
* @return {Function} middleware
*/

export default function favicon (resource = 'favicon.ico', options = {}) {
  const ONE_YEAR_MS = 60 * 60 * 24 * 365 * 1000 // 1 year

  let icon // favicon cache
  const maxAge = Number.isInteger(options.maxAge) ? Math.min(Math.max(0, options.maxAge), ONE_YEAR_MS) : ONE_YEAR_MS

  if (Buffer.isBuffer(resource)) {
    icon = createIcon(Buffer.from(resource), maxAge)
  } else if (typeof resource === 'string') {
    icon = createIcon(readFileSync(resolve(resource)), maxAge)
  } else {
    throw new TypeError('resource must be path to favicon or buffer')
  }

  return (req, res, next) => {
    if (req.path === '/favicon.ico') {
      if (req.method === 'GET' || req.method === 'HEAD') {
        send(req, res, icon)
      } else {
        res.statusCode = (req.method === 'OPTIONS') ? 200 : 405
        res.setHeader('Allow', 'GET, HEAD, OPTIONS')
        res.setHeader('Content-Length', '0')
        res.end()
      }
    } else {
      next()
    }
  }
}

/**
 * Create icon data from Buffer and max-age.
 *
 * @private
 * @param {Buffer} buf
 * @param {number} maxAge
 * @return {object}
 */

function createIcon (buf, maxAge) {
  return {
    body: buf,
    headers: {
      'Cache-Control': `public, max-age=${Math.floor(maxAge / 1000)}`,
      ETag: etag(buf)
    }
  }
}

/**
 * Determine if the cached representation is fresh.
 *
 * @param {object} req
 * @param {object} res
 * @return {boolean}
 * @private
 */

function isFresh (req, res) {
  return fresh(req.headers, {
    etag: res.getHeader('ETag'),
    'last-modified': res.getHeader('Last-Modified')
  })
}

/**
 * Send icon data in response to a request.
 *
 * @private
 * @param {IncomingMessage} req
 * @param {OutgoingMessage} res
 * @param {object} icon
 */

function send (req, res, icon) {
  // Set headers
  const headers = icon.headers
  const keys = Object.keys(headers)
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i]
    res.setHeader(key, headers[key])
  }

  // Validate freshness
  if (isFresh(req, res)) {
    res.statusCode = 304
    res.end()
  } else {
    // Send icon
    res.statusCode = 200
    res.setHeader('Content-Length', icon.body.length)
    res.setHeader('Content-Type', 'image/x-icon')
    res.end(icon.body)
  }
}
